git_toprepo.py

```diff
@@ -483,21 +490,29 @@ def extract_mapping(self, prefix: str) -> Dict[str, "ConfigDict"]:
         """Extracts for example submodule.<name>.<key>=<value>."""
         assert not prefix.endswith("."), prefix
         prefix += "."
-        ret = defaultdict(ConfigDict)
+        ret: DefaultDict[str, ConfigDict] = defaultdict(ConfigDict)
         for key, values in self.items():
             if key.startswith(prefix):
                 name, subkey = key[len(prefix) :].split(".", 1)
                 ret[name][subkey] = values
         return ret
 
     def get_singleton(
-        self, key, default: Optional[str] = _ConfigDict_unset
+        self, key: str, default: Optional[str] = _ConfigDict_unset
     ) -> Optional[str]:
-        """Varifies that there are no conflicting configuration values."""
+        """Verifies that there are no conflicting configuration values."""
         if default == _ConfigDict_unset:
             values = self[key]
         else:
-            values = self.get(key, [default])
+            maybe = self.get(key)
+            if maybe:
+                values = maybe
+            else:
+                if default:
+                    values = [default]
+                else:
+                    return None
+
```

Comment:
Returning `None` is wrong, `default` should be returned.
If `self[key]` is `[]`, `None` or `0`, that value should still be returned, not `default`.

Reply:
Ack, I will fork this out as this function may need some more thought. As I do not quite understand what we want here.
I agree that the if default check is needlessly sloppy and we should improve that. Though the typing of this function is complex to handle the type reflection/polymorphism.

Do you mean to just explicitly check against `is not None`? or a more advanced control flow?
- if default
+ if default is not None:

- return None
+ return Default

-------------------------------------------------------------
Here are some more questions that may or may not be relevant:

Does the config dict contain values: Union[str, int, List[...]]?  I wrote the code assuming that it was only Lists of strings
> ConfigDict(DefaultDict[str, List[str]])

I do not know how to assign List[Optional[str]] into the `values` sequence without issues as it is used later on.

> 1. mypy: Value of type variable "SupportsRichComparisonT" of "sorted" cannot be "str | None" [type-var]

Reply:
Let's solve this in a later PR.